[Inductor] deterministic mode

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164533
* #164532
* #164525
* __->__ #163589


Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.
- benchmarking for computation communication reordering pass

The mode definitely has perf hit.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

shunting314 committed:
[Inductor] deterministic mode

[ghstack-poisoned]

eellison reviewed (commented):
Looks good! lets get takes from other folks too before landing. this doesnt handle interaction with caching, but that can be done subsequently, I know @nmacchioni has some early starts there.

PaulZhang12 commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2807,6 +2824,15 @@ def adapt_config_for_tiling(
     )
 
 
+def filter_reduction_configs_for_determinism(configs: list[Config]) -> list[Config]:
```
configs[0] case here should already pick quite good configs given the updated heuristics, can you maybe retry also on top of https://github.com/pytorch/pytorch/pull/162447 ? I'm curious where the gap is

shunting314 replied:
Changed this part a bit by:
1. do autotuning for xblock
2. pick a group with the smallest rblock

Now, the perf in deterministic mode is 1.416x .  This may be further improved if we consider features like xnumel/rnumel/num-load/num-store/num-contiguous-acess/num-non-contiguous-access etc. 

I think picking configs[0] mainly not cover the case that a reduction have both contiguous and non-contiguous access

v0i0 replied:
i am not sure this is sufficient for non-determinism, specifically when rblock is small, have you checked? i'd think that tuning xblock can change the # of elements per thread, which might impact how the reduction is performed?

shunting314 replied:
> i am not sure this is sufficient for non-determinism, specifically when rblock is small, have you checked?

Why it's insufficient? We do step 2 with deterministic heuristics. In this version, I just return the one with smallest RBLOCK.

v0i0 replied:
ah got it

shunting314 replied:
I collected the autotuning results for huggingface benchmark suite ( https://gist.github.com/shunting314/a52aa2b436229d1b92d6d116163d1a4a ) and analyze manually 5 of the models (https://gist.github.com/shunting314/c9322d90ce405ec0e24aac3f97eb1963 ) to come up with some heuristics.  These heuristics make BertForMaskedLM having on par perf with or without filtering.

shunting314 replied:
The heuristic works even better than I think. I ran a perf test forcing TORCHINDUCTOR_FORCE_FILTER_REDUCTION_CONFIGS to be true (i.e. do reduction config filtering. But mm-padding etc are still active). [result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2025%20Sep%202025%2000%3A24%3A47%20GMT&stopTime=Thu%2C%2002%20Oct%202025%2000%3A24%3A47%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=gh/shunting314/228/head&lCommit=582267e91d65739a43cd77434746bbdf31e059c6&rBranch=main&rCommit=77b9aac6c2ab41e07ca2643260ca3e6f059abc4a).

TB: 3% perf hit, 3s compilation time speedup
HF: 6% perf hit, 1s compilation time speedup
TIMM: 1% perf hit, 6s compilation time speedup.

Looks like this is very beneficial for TIMM.

<img width="1488" height="584" alt="Screenshot 2025-10-01 at 5 26 17 PM" src="https://github.com/user-attachments/assets/452e09a1-95dd-4fe9-a43b-da46447b6ce2" />

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. Not sure about XBLOCK. Skip the tuning for all of them for now.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

v0i0 commented on torch/_inductor/config.py:
```diff
@@ -700,6 +700,10 @@ def use_autoheuristic(name: str) -> bool:
 
+# A deterministic mode that skips any on device benchmarking in Inductor
+# if we know they affect numerics.  WARNING: Expect perf hit in this mode.
+deterministic = os.getenv("TORCHINDUCTOR_DETERMINISTIC") == "1"
```
for my understanding (and maybe to elaborate the comment): this flag is intended as a kind of override, right? i.e. it takes precedence over all other config options that might trigger benchmark-guided optimization?

shunting314 replied:
yes

v0i0 reviewed (approved):
grepping through the code suggests some other configs as well (e.g. runtime_estimations_mms_benchmark) that should be overridden?

eellison commented:
> grepping through the code suggests some other configs as well (e.g. runtime_estimations_mms_benchmark) that should be overridden?

Would it be safer to make do_bench error when determinism is set ? And we can explicitly allow it in portions of code we have vetted as equivalent.

shunting314 commented:
> Would it be safer to make do_bench error when determinism is set ? And we can explicitly allow it in portions of code we have vetted as equivalent.

make sense. This would also guard against benchmarking code added in future as long as people do benchmarking with the small set of benchmarking functions in Inductor. I'll add this in a stacked PR

shunting314 commented:
> Would it be safer to make do_bench error when determinism is set ? And we can explicitly allow it in portions of code we have vetted as equivalent.

hmm, people call triton.testing.do_bench directly sometimes. So even if we add this, it would still be best effect but can not always guaranteed no benchmarking happening.

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

drisspg commented on torch/_inductor/select_algorithm.py:
```diff
@@ -2600,6 +2600,16 @@ def cache_clear(self) -> None:
+        if len(externs) > 0:
+            return externs[0]
+        else:
+            return choices[0]
```
So by default Flex only picks 1 choice and then we dump them all for max-autotune. That choice is forsure not always the best but I dont think this will be too much of a delta unless max-autotune is enabled

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

nmacchioni commented:
> @nmacchioni
> 
> > the cache will spit out an override file that users can copy/paste to replay the same results (and I guess this is where folks can choose to meddle with the results)
> 
> Is there a way to force determinism and error if the override file doesn't have enough information?

Yeah, in my mind we can either throw an exception if the override file doesn't have the required information to be deterministic or we can fallback to synchronizing off of the remote db. OSS will have to throw an exception by default, as we're not providing the remote db there. Internally we'll just have a toggle for whatever behavior is wanted

shunting314 committed:
Update on "[Inductor] deterministic mode"



Add a deterministic mode to skip the on device benchmarking that we know should affect numeric. This include
- pad-mm
- dynamic rblock scaling
- template autotuning
- coordinate descent tuning for reduction
- reduction config autotuning in CachingAutotuner.  For reduction both RBLOCK, num_warps should affect numeric. XBLOCK does not. We can still autotune XBLOCK for reductions.
- benchmarking for computation communication reordering pass

The mode definitely has perf hit.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]
